test/test/CompilerTest.scala

```diff
@@ -140,54 +153,69 @@ abstract class CompilerTest {
     * compileFile resp. compileDir. */
-  def compileFiles(path: String, args: List[String] = Nil, verbose: Boolean = true, runTest: Boolean = false)
-      (implicit defaultOptions: List[String]): Unit = {
+  def compileFiles(path: String, args: List[String] = Nil, verbose: Boolean = true, runTest: Boolean = false,
+                   negTest: Boolean = false, compileSubDirs: Boolean = true)(implicit defaultOptions: List[String]): Unit = {
     val dir = Directory(path)
     val fileNames = dir.files.toArray.map(_.jfile.getName).filter(name => (name endsWith ".scala") || (name endsWith ".java"))
     for (name <- fileNames) {
       if (verbose) log(s"testing $path$name")
-      compileFile(path, name, args, 0, "", runTest)
-    }
-    for (subdir <- dir.dirs) {
-      if (verbose) log(s"testing $subdir")
-      compileDir(path, subdir.jfile.getName, args, 0, runTest)
+      compileFile(path, name, args, negTest, "", runTest)
     }
+    if (compileSubDirs)
+      for (subdir <- dir.dirs) {
+        if (verbose) log(s"testing $subdir")
+        compileDir(path, subdir.jfile.getName, args, negTest, runTest)
+      }
   }
   def runFiles(path: String, args: List[String] = Nil, verbose: Boolean = true)
       (implicit defaultOptions: List[String]): Unit =
     compileFiles(path, args, verbose, true)
 
   /** Compiles the given list of code files. */
-  def compileList(testName: String, files: List[String], args: List[String] = Nil, xerrors: Int = 0)
+  def compileList(testName: String, files: List[String], args: List[String] = Nil, negTest: Boolean = false)
       (implicit defaultOptions: List[String]): Unit = {
-    if (!generatePartestFiles || !partestableList(testName, files, args ++ defaultOptions, xerrors)) {
-      compileArgs((files ++ args).toArray, xerrors)
+    if (!generatePartestFiles || !partestableList(testName, files, args ++ defaultOptions)) {
+      val expErrors = expectedErrors(files, negTest)
+      compileArgs((files ++ args).toArray, expErrors, negTest)
     } else {
       val destDir = Directory(DPConfig.testRoot + JFile.separator + testName)
       files.foreach({ file =>
         val jfile = new JFile(file)
         recCopyFiles(jfile, destDir / jfile.getName)
       })
-      compileDir(DPConfig.testRoot + JFile.separator, testName, args, xerrors)
+      compileDir(DPConfig.testRoot + JFile.separator, testName, args, negTest)
       destDir.deleteRecursively
     }
   }
 
   // ========== HELPERS =============
 
-  private def compileArgs(args: Array[String], xerrors: Int = 0)
+  private def expectedErrors(filePaths: List[String], negTest: Boolean): List[ErrorsInFile] = {
+    if (negTest) {
+      filePaths.map(getErrors(_))
+    } else Nil
+  }
+
+  private def expectedErrors(filePath: String, negTest: Boolean): List[ErrorsInFile] = expectedErrors(List(filePath), negTest)
+
+  private def compileArgs(args: Array[String], expectedErrorsPerFile: List[ErrorsInFile], negTest: Boolean)
```

Comment:
isn't it always true that `negTest == expectedErrorsPerFile.nonEmpty`?
If yes, I propose to remove negTest from argument list and give `expectedErrorsPerFile` default value of `Nil`.